<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="Huwenbo Shi">
        <link rel="canonical" href="https://huwenboshi.github.io/s-ldxr/gcorsq/">
        <link rel="shortcut icon" href="../img/favicon.ico">
        <title>Estimating enrichment - S-LDXR</title>
        <link href="../css/bootstrap.min.css" rel="stylesheet">
        <link href="../css/font-awesome.min.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">

        <script src="../js/jquery-1.10.2.min.js" defer></script>
        <script src="../js/bootstrap.min.js" defer></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
        <script>hljs.initHighlightingOnLoad();</script>
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-XXXXXXXX-X', 'auto');
            ga('send', 'pageview');
        </script> 
    </head>

    <body>
        <div class="navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
            <div class="container">
                <a class="navbar-brand" href="..">S-LDXR</a>
                <!-- Expander button -->
                <button type="button" class="navbar-toggler" data-toggle="collapse" data-target="#navbar-collapse">
                    <span class="navbar-toggler-icon"></span>
                </button>

                <!-- Expanded navigation -->
                <div id="navbar-collapse" class="navbar-collapse collapse">
                        <!-- Main navigation -->
                        <ul class="nav navbar-nav">
                            <li class="navitem">
                                <a href=".." class="nav-link">Home</a>
                            </li>
                            <li class="navitem">
                                <a href="../input_format/" class="nav-link">Input Format</a>
                            </li>
                            <li class="dropdown active">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">Estimation <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../ldscore/" class="dropdown-item">Estimating LD score</a>
</li>
                                    
<li>
    <a href="./" class="dropdown-item active">Estimating enrichment</a>
</li>
                                </ul>
                            </li>
                            <li class="navitem">
                                <a href="../faq/" class="nav-link">FAQ</a>
                            </li>
                        </ul>

                    <ul class="nav navbar-nav ml-auto">
                        <li class="nav-item">
                            <a href="#" class="nav-link" data-toggle="modal" data-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li class="nav-item">
                                <a rel="prev" href="../ldscore/" class="nav-link">
                                    <i class="fa fa-arrow-left"></i> Previous
                                </a>
                            </li>
                            <li class="nav-item">
                                <a rel="next" href="../faq/" class="nav-link">
                                    Next <i class="fa fa-arrow-right"></i>
                                </a>
                            </li>
                            <li class="nav-item">
                                <a href="https://github.com/huwenboshi/s-ldxr/edit/master/docs/gcorsq.md" class="nav-link"><i class="fa fa-github"></i> Edit on GitHub</a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
            <div class="row">
                    <div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
    <div class="navbar-header">
        <button type="button" class="navbar-toggler collapsed" data-toggle="collapse" data-target="#toc-collapse" title="Table of Contents">
            <span class="fa fa-angle-down"></span>
        </button>
    </div>

    
    <div id="toc-collapse" class="navbar-collapse collapse card bg-secondary">
        <ul class="nav flex-column">
            
            <li class="nav-item" data-level="1"><a href="#estimating-enrichment-of-stratified-squared-trans-ethnic-genetic-correlation" class="nav-link">Estimating enrichment of stratified squared trans-ethnic genetic correlation</a>
              <ul class="nav flex-column">
              </ul>
            </li>
        </ul>
    </div>
</div></div>
                    <div class="col-md-9" role="main">

<h1 id="estimating-enrichment-of-stratified-squared-trans-ethnic-genetic-correlation">Estimating enrichment of stratified squared trans-ethnic genetic correlation</h1>
<p>This page describes the steps to estimate enrichment of stratified squared
trans-ethnic genetic correlation,
\(\lambda^2(C) ={ {r^2_g(C)} \over {r^2_g} }\), the ratio between squared
trans-ethnic genetic correlation of annotation \( C \) and genome-wide
squared trans-ethnic genetic correlation.</p>
<h3 id="typical-command">Typical command</h3>
<p>S-LDXR estimates \(r^2_{g}(C)\) and \(\lambda^2(C)\) with the
following command.</p>
<div class="codehilite"><pre><span></span><code><span class="err">python &lt;software directory&gt;/s-ldxr.py \</span>
<span class="err">    --gcor &lt;summary stats directory for EAS&gt;/EAS_sumstats.gz \</span>
<span class="err">           &lt;summary stats directory for EUR&gt;/EUR_sumstats.gz \</span>
<span class="err">    --ref-ld-chr &lt;baseline LD score directory&gt;/EAS_EUR_baseline_chr \</span>
<span class="err">                 &lt;AVGLLD LD score directory&gt;/EAS_EUR_avglld_chr \</span>
<span class="err">                 &lt;BSTAT LD score directory&gt;/EAS_EUR_bstat_chr \</span>
<span class="err">                 &lt;ALLELEAGE LD score directory&gt;/EAS_EUR_alleleage_chr \</span>
<span class="err">    --w-ld-chr &lt;regression weight directory&gt;/EAS_EUR_weight_chr \</span>
<span class="err">    --frqfile &lt;EAS MAF directory&gt;/1000G.EAS. \</span>
<span class="err">              &lt;EUR MAF directory&gt;/1000G.EUR. \</span>
<span class="err">    --annot &lt;baseline annotation directory&gt;/baseline. \</span>
<span class="err">            &lt;AVGLLD annotation directory&gt;/avglld. \</span>
<span class="err">            &lt;BSTAT annotation directory&gt;/bstat. \</span>
<span class="err">            &lt;ALLELEAGE annotation directory&gt;/alleleage. \</span>
<span class="err">    --apply-shrinkage 0.5 \</span>
<span class="err">    --save-pseudo-coef \</span>
<span class="err">    --out TRAIT_EAS_EUR.txt</span>
</code></pre></div>


<p>This command typically takes 10 to 15 minutes to run on a stand alone computer.</p>
<p>Here are the meanings of the flags:</p>
<ul>
<li>
<p><code>--gcor</code> specifies the summary stats files. This flag takes 2 arguments -
summary stats for population 1 and summary stats for population 2.</p>
</li>
<li>
<p><code>--ref-ld-chr</code> specifies prefix of the LD score files. This flag takes one
or more arguments â€“ one may put as many LD score files as one wishes.</p>
</li>
<li>
<p><code>--w-ld-chr</code> specifies prefix of the regression weights. These are
standardized LD scores calculated from regression SNPs.</p>
</li>
<li>
<p><code>--frqfile</code> specifies prefix of minor allele frequency files.</p>
</li>
<li>
<p><code>--annot</code> specifies prefix of the annotation files. This flags also takes
one or more arguments.</p>
</li>
</ul>
<div style="background-color:rgba(230, 230, 250, 1.0);">
<b>Note</b>: The order one specifies the annotation files must
be the same as the order one specifies the LD score files. The annotation
files must also be the same files that one uses to obtain the LD scores.
</div>

<ul>
<li>
<p><code>--apply-shrinkage</code> adjusts the level of shrinkage (the \(\alpha\) tuning
parameter in the paper). This should be a number between 0 and 1.</p>
</li>
<li>
<p><code>--save-pseudo-coef</code> If this flag is specified, jackknife pseudo values of
the coefficients will be saved. This flag is optional.</p>
</li>
<li>
<p><code>--out</code> specifies the output file name.</p>
</li>
</ul>
<h3 id="output">Output</h3>
<p>After executing the above command, 5 files will be generated.</p>
<ul>
<li>
<p><code>TRAIT_EAS_EUR.txt</code> output file containing the estimates.</p>
</li>
<li>
<p><code>TRAIT_EAS_EUR.txt.log</code> log file containing information for debugging.</p>
</li>
<li>
<p><code>TRAIT_EAS_EUR.txt.pseudo_tau1.gz</code> jackknife pseudo values for \(\tau_C\)
coefficients for population 1.</p>
</li>
<li>
<p><code>TRAIT_EAS_EUR.txt.pseudo_tau2.gz</code> jackknife pseudo values for \(\tau_C\)
coefficients for population 2.</p>
</li>
<li>
<p><code>TRAIT_EAS_EUR.txt.pseudo_theta.gz</code> jackknife pseudo values for
\(\theta_C\) genetic covariance coefficients.</p>
</li>
</ul>
<div style="background-color:rgba(230, 230, 250, 1.0);">
<b>Note</b>: The pseudo coefficients will only be saved if the
--save-pseudo-coef flag is specified.
</div>

<h3 id="continuous-valued-annotations">Continuous-valued annotations</h3>
<p>The following command estimates enrichment of stratified squared trans-ethnic
genetic correlation of quintiles of continuous-valued annotations.</p>
<div class="codehilite"><pre><span></span><code><span class="err">python &lt;software directory&gt;/cont_annot_gcor.py \</span>
<span class="err">    --coef TRAIT_EAS_EUR.txt \</span>
<span class="err">    --frqfile &lt;EAS MAF directory&gt;/1000G.EAS. \</span>
<span class="err">              &lt;EUR MAF directory&gt;/1000G.EUR. \</span>
<span class="err">    --annot &lt;baseline annotation directory&gt;/baseline. \</span>
<span class="err">            &lt;AVGLLD annotation directory&gt;/avglld. \</span>
<span class="err">            &lt;BSTAT annotation directory&gt;/bstat. \</span>
<span class="err">            &lt;ALLELEAGE annotation directory&gt;/alleleage. \</span>
<span class="err">    --names AVGLLD BSTAT ALLELEAGE \</span>
<span class="err">    --nbins 5 \</span>
<span class="err">    --out TRAIT_EAS_EUR_contannot.txt</span>
</code></pre></div>


<p>This step typically takes 2 to 5 minutes to run on a stand alone computer.</p>
<p>Here are the meanings of the flags.</p>
<ul>
<li>
<p><code>--coef</code> specifies the output from the previous step. The jackknife pseudo
coefficients will be loaded automatically.</p>
</li>
<li>
<p><code>--frqfile</code> specifies prefix of minor allele frequency files.</p>
</li>
<li>
<p><code>--annot</code> specifies prefix of the annotation files. This flags also takes
one or more arguments.</p>
</li>
</ul>
<div style="background-color:rgba(230, 230, 250, 1.0);">
<b>Note</b>: The order one specifies the annotation files must be the same as the
order of annotations in TRAIT_EAS_EUR.txt.
</div>

<ul>
<li>
<p><code>--names</code> specifies the names of the continuous annotations for which one
wishes to compute enrichment at quintiles.</p>
</li>
<li>
<p><code>--nbins</code> specifies the number of bins to bin the SNPs based on the values
of their continuous annotation. The default is 5 (i.e. quintiles).</p>
</li>
<li>
<p><code>--out</code> specifies the output file name.</p>
</li>
</ul>
<p>Additionally, users may use the <code>--apply-shrinkage</code> flag to adjust the level
of shrinkage.</p>
<p>After executing the above command, 2 files will be created.</p>
<ul>
<li>
<p><code>TRAIT_EAS_EUR_contannot.txt</code> contains the estimates.</p>
</li>
<li>
<p><code>TRAIT_EAS_EUR_contannot.txt.log</code> is the log file for debugging purpose.</p>
</li>
</ul>
<h3 id="expected-r2_gc-and-lambda2c-from-continuous-valued-annotations">Expected \(r^2_g(C)\) and \(\lambda^2(C)\) from continuous-valued annotations</h3>
<p>Estimating expected \(r^2_g(C)\) and \(\lambda^2(C)\) from
continuous-valued annotations requires two steps.</p>
<p>The first step gets the coefficients (\(\tau_{1C}\), \(\tau_{2C}\),
and \(\theta_{C}\)) of each continuous-valued annotations</p>
<div class="codehilite"><pre><span></span><code><span class="err">python &lt;software directory&gt;/s-ldxr.py \</span>
<span class="err">    --gcor &lt;summary stats directory for EAS&gt;/EAS_sumstats.gz \</span>
<span class="err">           &lt;summary stats directory for EUR&gt;/EUR_sumstats.gz \</span>
<span class="err">    --ref-ld-chr &lt;base LD score directory&gt;/EAS_EUR_allelic_chr \</span>
<span class="err">                 &lt;AVGLLD LD score directory&gt;/EAS_EUR_allelic_chr \</span>
<span class="err">                 &lt;BSTAT LD score directory&gt;/EAS_EUR_allelic_chr \</span>
<span class="err">                 &lt;ALLELEAGE LD score directory&gt;/EAS_EUR_allelic_chr \</span>
<span class="err">    --w-ld-chr &lt;regression weight directory&gt;/EAS_EUR_weight_chr \</span>
<span class="err">    --frqfile &lt;EAS MAF directory&gt;/1000G.EAS. \</span>
<span class="err">              &lt;EUR MAF directory&gt;/1000G.EUR. \</span>
<span class="err">    --annot &lt;base annotation directory&gt;/base. \</span>
<span class="err">            &lt;AVGLLD annotation directory&gt;/avglld. \</span>
<span class="err">            &lt;BSTAT annotation directory&gt;/bstat. \</span>
<span class="err">            &lt;ALLELEAGE annotation directory&gt;/alleleage. \</span>
<span class="err">    --save-pseudo-coef \</span>
<span class="err">    --out ./TRAIT_EAS_EUR_step1.txt</span>
</code></pre></div>


<p>This command typically takes 2 to 5 minutes to run on a stand alone computer.</p>
<div style="background-color:rgba(230, 230, 250, 1.0);">
<b>Note</b>: It is important to always include the base (not
baseline) annotation.
</div>

<p>The output is the same as that of a typical command.</p>
<p>The second step uses the coefficients from the first step to obtain expected
\(r^2_g(C)\) and \(\lambda^2(C)\) from continuous-valued annotations.</p>
<div class="codehilite"><pre><span></span><code><span class="err">python &lt;software directory&gt;/pred_binannot_from_contannot.py \</span>
<span class="err">    --coef ./TRAIT_EAS_EUR_step1.txt \</span>
<span class="err">    --frqfile &lt;EAS MAF directory&gt;/1000G.EAS. \</span>
<span class="err">              &lt;EUR MAF directory&gt;/1000G.EUR. \</span>
<span class="err">    --cont-annot &lt;base annotation directory&gt;/base. \</span>
<span class="err">                 &lt;AVGLLD annotation directory&gt;/avglld. \</span>
<span class="err">                 &lt;BSTAT annotation directory&gt;/bstat. \</span>
<span class="err">                 &lt;ALLELEAGE annotation directory&gt;/alleleage. \</span>
<span class="err">    --bin-annot &lt;base annotation directory&gt;/base. \</span>
<span class="err">                &lt;binary annotation directory&gt;/annot_name. \</span>
<span class="err">    --apply-shrinkage 0.5 \</span>
<span class="err">    --out ./TRAIT_EAS_EUR_step2.txt</span>
</code></pre></div>


<p>This command typically takes 2 to 5 minutes to run on a stand alone computer.</p>
<div style="background-color:rgba(230, 230, 250, 1.0);">
<b>Note</b>: It is important to always include the base (not
baseline) annotation.
</div>

<p>The output is the same as that of the command for continuous-valued
annotations.</p>
<h3 id="interpreting-the-output">Interpreting the output</h3>
<p>The output files of S-LDXR contain the following columns.</p>
<ol>
<li>
<p><code>ANNOT</code> name of the annotation</p>
</li>
<li>
<p><code>NSNP</code> number of SNPs for binary annotations (sum of annotation values for
continuous-valued annotations)</p>
</li>
<li>
<p><code>STD</code> standard deviation of the annotation across SNPs</p>
</li>
<li>
<p><code>TAU1</code> heritability coefficient of population 1</p>
</li>
<li>
<p><code>TAU1_SE</code> standard error heritability coefficient of population 1</p>
</li>
<li>
<p><code>TAU2</code> heritability coefficient of population 2</p>
</li>
<li>
<p><code>TAU2_SE</code> standard error heritability coefficient of population 2</p>
</li>
<li>
<p><code>THETA</code> trans-ethnic genetic covariance coefficient</p>
</li>
<li>
<p><code>THETA_SE</code> standard error of trans-ethnic genetic covariance coefficient</p>
</li>
<li>
<p><code>HSQ1</code> heritability in population 1</p>
</li>
<li>
<p><code>HSQ1_SE</code> standard error of heritability in population 1</p>
</li>
<li>
<p><code>HSQ2</code> heritability in population 2</p>
</li>
<li>
<p><code>HSQ2_SE</code> standard error of heritability in population 2</p>
</li>
<li>
<p><code>GCOV</code> trans-ethnic genetic covariance</p>
</li>
<li>
<p><code>GCOV_SE</code> standard error of trans-ethnic genetic covariance</p>
</li>
<li>
<p><code>GCORSQ</code> stratified squared trans-ethnic genetic correlation</p>
</li>
<li>
<p><code>GCORSQ_SE</code> standard error of stratified squared trans-ethnic genetic
correlation </p>
</li>
<li>
<p><code>HSQ1_ENRICHMENT</code> heritability enrichment in population 1</p>
</li>
<li>
<p><code>HSQ1_ENRICHMENT_SE</code> standard error of heritability enrichment in
population 1</p>
</li>
<li>
<p><code>HSQ2_ENRICHMENT</code> heritability enrichment in population 2</p>
</li>
<li>
<p><code>HSQ2_ENRICHMENT_SE</code> standard error of heritability enrichment in
population 2</p>
</li>
<li>
<p><code>GCOV_ENRICHMENT</code> genetic covariance enrichment</p>
</li>
<li>
<p><code>GCOV_ENRICHMENT_SE</code> standard error of genetic covariance enrichment</p>
</li>
<li>
<p><code>GCORSQ_ENRICHMENT</code> estimated enrichment of stratified squared trans-ethnic
genetic correlation enrichment</p>
</li>
<li>
<p><code>GCORSQ_ENRICHMENT_SE</code> standard error of estimated enrichment of stratified
squared trans-ethnic genetic correlation</p>
</li>
</ol></div>
            </div>
        </div>

        <footer class="col-md-12">
            <hr>
                <p>Copyright &copy; 2019 - 2024 Huwenbo Shi</p>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="../js/base.js" defer></script>
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML" defer></script>
        <script src="../assets/mathjaxhelper.js" defer></script>
        <script src="../search/main.js" defer></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form>
                    <div class="form-group">
                        <input type="search" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
